File: ruby/robot-simulator/robot_simulator.rb
class Robot

  attr_reader :bearing, :coordinates
  DIRECTIONS = [:west, :north, :east, :south].freeze

  def orient(direction)
    raise ArgumentError unless DIRECTIONS.include? direction
    @bearing = direction
  end

  [:right, :left].each do |direction|
    define_method("turn_#{direction}") do
      @bearing = direction == :right ? DIRECTIONS[(DIRECTIONS.index(@bearing) + 1)%4]
                                     : DIRECTIONS[(DIRECTIONS.index(@bearing) - 1)%4]
    end
  end

  def at(x, y)
    @coordinates = [x, y]
  end

  def advance
    bearing_index = DIRECTIONS.index(@bearing)
    bearing_index.even? ? @coordinates[0] = @coordinates[0] + (bearing_index-1)
                        : @coordinates[1] = @coordinates[1] - (bearing_index-2)
  end

end

class Simulator

  INSTRUCTIONS = {
    'L' => :turn_left,
    'R' => :turn_right,
    'A' => :advance
  }.freeze

  def instructions(commands)
    commands.chars.map { |e| INSTRUCTIONS[e] }
  end

  def place(robot, x:, y:, direction:)
    robot.at(x, y)
    robot.orient(direction)
  end

  def evaluate(robot, commands)
    instructions(commands).each do |cmnd|
      robot.send(cmnd)
    end
  end

end

File: python/robot-simulator/robot_simulator.py
# Globals for the bearings
# Change the values as you see fit
NORTH = 0
EAST = 1
SOUTH = 2
WEST = 3

class Robot(object):
    def __init__(self, bearing=NORTH, x=0, y=0):
        self.bearing = bearing
        self.coordinates = (x, y)

    def turn_right(self):
        self.bearing = (self.bearing + 1) % 4

    def turn_left(self):
        self.bearing = (self.bearing - 1) % 4

    def advance(self):
        if self.bearing == NORTH or self.bearing == SOUTH:
            self.coordinates = (self.coordinates[0], self.coordinates[1] + 1 - self.bearing)
        else:
            self.coordinates = (self.coordinates[0] + 2 - self.bearing, self.coordinates[1])

    def simulate(self, commands):
        for command in commands:
            if command == 'R':
                self.turn_right()
            elif command == 'L':
                self.turn_left()
            elif command == 'A':
                self.advance()
    
